shared/src/main/scala/mlscript/Typer.scala

```diff
@@ -1021,6 +1021,70 @@ class Typer(var dbg: Boolean, var verbose: Bool, var explainErrors: Bool, val ne
         val desug = If(IfThen(lhs, rhs), S(Var("false")))
         term.desugaredTerm = S(desug)
         typeTerm(desug)
+      case App(f: Term, a @ Tup(fields)) if (fields.exists(x => x._1.isDefined)) =>
+        val f_ty = typeTerm(f)
+        val fun_ty: SimpleType = f_ty.unwrapProxies match {
+          case tv: TypeVariable =>
+            def getLowerboundFuns(tv: TypeVariable): List[FunctionType] = {
+              val res: List[FunctionType] = tv.lowerBounds.map(x => 
+                x.unwrapProvs match {
+                  case PolymorphicType(_, AliasOf(fun_ty @ FunctionType(_, _))) =>
+                    List(fun_ty)
+                  case tvv: TypeVariable =>
+                    getLowerboundFuns(tvv) 
+                  case _ =>
+                    err("match error", f.toLoc)
+                    Nil
+                }).flatten
+              res
+            }
+            val funs = getLowerboundFuns(tv)
+            funs match {
+              case x :: Nil => 
+                funs.head
+              case Nil =>
+                err("Cannot extract any function", f.toLoc)
+              case _ =>
+                err(s"More than one fun type found: ${funs}", f.toLoc)
+            }
+          case PolymorphicType(_, AliasOf(fun_ty @ FunctionType(_, _))) =>
+            fun_ty
+          case FunctionType(_, _) =>
+            f_ty
+          case _ =>
+            err("match error", f.toLoc)
+            f_ty
+        }
+        // check if all
+        val hasUntypedArg = fun_ty.unwrapProxies match {
+          case FunctionType(TupleType(fields), _) =>
+            fields.exists(x => x._1 match {
+              case Some(arg) =>
+                false
+              case N =>
+                true
+            })
+          case _ => 
+            err("match error", f.toLoc)
+            true
+        }
+        if (hasUntypedArg) {
+          err("Cannot use named args if there is untyped argument", a.toLoc)
+        } else {
+          val argsList = fun_ty.unwrapProxies match {
+            case FunctionType(TupleType(fields), _) =>
+              fields.map(x => x._1 match {
+                case Some(arg) =>
+                  arg
+                case N =>
+                  Var("error")
```

Comment:
Add a comment here to explain that this is used as a dummy term referring to the `error` builtin.